### PR TITLE
feat: Phase 4.1 - Add flat output mode support to convertPost method

### LINKED ISSUE #51: 4.1 Update Converter convertPost Method for Flat Mode
### Overview
Update the `convertPost` method to support flat output mode by reading `outputStructure` from options and routing to the appropriate processing methods.

### Tasks
- [x] Read `outputStructure` from options (default to nested mode)
- [x] Calculate image directory based on mode:
  - Nested: `{output}/{slug}/`
  - Flat: `{output}/../{imageFolderName}/`
- [x] Create image directory before processing
- [x] Use `processWithContext()` for flat mode
- [x] Use existing `process()` for nested mode (backwards compatible)
- [x] Create FileWriter with appropriate `outputMode`

### Proposed Changes

**File: `src/converter.ts`**

Update `convertPost` method (around line 375-455):

```typescript
async convertPost(
  post: HashnodePost,
  outputDir: string,
  options?: ConversionOptions
): Promise<ConvertedPost> {
  const slug = this.extractSlugSafely(post, 0);
  const outputStructure = options?.outputStructure ?? { mode: 'nested' };
  const isFlat = outputStructure.mode === 'flat';

  try {
    // Step 1: Parse post metadata
    const metadata = this.postParser.parse(post);

    // Step 2: Transform markdown
    const transformedMarkdown = this.markdownTransformer.transform(
      metadata.contentMarkdown
    );

    // Step 3: Determine image directory and prepare it
    let imageDir: string;
    let imagePathPrefix: string;

    if (isFlat) {
      // Flat mode: images go to sibling folder
      const parentDir = path.dirname(outputDir);
      const imageFolderName = outputStructure.imageFolderName ?? '_images';
      imageDir = path.join(parentDir, imageFolderName);
      imagePathPrefix = outputStructure.imagePathPrefix ?? '/images';
    } else {
      // Nested mode: images go into post directory
      imageDir = path.join(outputDir, metadata.slug);
      imagePathPrefix = '.';
    }

    // Create image directory if needed
    if (!fs.existsSync(imageDir)) {
      fs.mkdirSync(imageDir, { recursive: true });
    }

    // Step 4: Process images
    // Note: Creating a new ImageProcessor with custom downloadOptions is safe because
    // download state is persisted via .downloaded-markers/ files on disk, not in-memory.
    // A new instance will read existing markers and skip already-downloaded images.
    // Custom options only affect retry behavior for new/failed downloads.
    const imageProcessor = options?.downloadOptions
      ? new ImageProcessor(options.downloadOptions)
      : this.imageProcessor;

    let imageResult: ImageProcessingResult;
    if (isFlat) {
      // Use context-aware processing for flat mode
      imageResult = await imageProcessor.processWithContext(transformedMarkdown, {
        imageDir,
        imagePathPrefix,
      });
    } else {
      // Use existing method for nested mode (backwards compatible)
      imageResult = await imageProcessor.process(transformedMarkdown, imageDir);
    }

    // Emit image events and track 403 errors
    this.emitImageDownloadedEvents(imageResult, metadata.slug);
    this.trackHttp403Errors(imageResult.errors, metadata.slug);

    // Step 5: Generate frontmatter
    const frontmatter = this.frontmatterGenerator.generate(metadata);

    // Step 6: Write file (FileWriter handles flat vs nested)
    const fileWriter = isFlat
      ? new FileWriter({ outputMode: 'flat' })
      : this.fileWriter;

    const outputPath = await fileWriter.writePost(
      outputDir,
      metadata.slug,
      frontmatter,
      imageResult.markdown
    );

    // Emit completion event
    this.emit('conversion-completed', {
      slug: metadata.slug,
      title: metadata.title,
      outputPath,
    });

    return {
      slug: metadata.slug,
      title: metadata.title,
      outputPath,
      success: true,
    };
  } catch (error) {
    // ... existing error handling unchanged
  }
}
```

### Data Flow

**Nested Mode:**
```
outputDir: src/_posts
imageDir: src/_posts/{slug}/
imagePathPrefix: .
output: src/_posts/{slug}/index.md
```

**Flat Mode:**
```
outputDir: src/_posts
imageDir: src/_images/  (sibling)
imagePathPrefix: /images
output: src/_posts/{slug}.md
```

### References
- Implementation Plan: [docs/IMPLEMENTATION_FLAT.md](docs/IMPLEMENTATION_FLAT.md)
- Depends on: Phase 2 and Phase 3 completion

### PR DESCRIPTION
**Pull request by alvincrespo**
## Phase 4, Step 4.1: Flat Output Mode Support

Implements flat output mode support in the `Converter.convertPost` method as specified in Phase 4, Step 4.1 of the flat output mode implementation plan.

**Related Issue**: Closes #51

---

## Summary

This PR adds support for flat output mode to the `convertPost` method, enabling the converter to write posts as `{slug}.md` files with images in a shared sibling directory (`_images`), instead of the current nested `{slug}/index.md` structure.

### Core Implementation ([src/converter.ts](src/converter.ts))

1. **Extract OutputStructure Configuration**
   - Read `outputStructure` from options with nested mode as default (backward compatible)
   - Early mode detection establishes single source of truth for conditional logic

2. **Calculate Image Directory and Path Prefix**
   - **Nested mode** (default): `imageDir = {output}/{slug}`, `imagePathPrefix = '.'`
   - **Flat mode**: `imageDir = {output}/../_images`, `imagePathPrefix = '/images'`
   - Support custom `imageFolderName` and `imagePathPrefix` options

3. **Route to Appropriate ImageProcessor Method**
   - Flat mode uses `processWithContext()` with custom paths
   - Nested mode uses existing `process()` method (unchanged)
   - Added import for `ImageProcessingResult` type

4. **Create FileWriter with Correct OutputMode**
   - Flat mode creates new `FileWriter({ outputMode: 'flat' })`
   - Nested mode reuses `this.fileWriter` for efficiency

5. **Enhanced Error Handling**
   - Added 'Image directory does not exist' to write error classification
   - Properly handles flat-mode-specific error scenarios

### Testing ([tests/integration/converter.test.ts](tests/integration/converter.test.ts))

Added 7 new integration tests covering:
- ✅ Basic flat mode usage with `processWithContext`
- ✅ Custom `imageFolderName` option
- ✅ Custom `imagePathPrefix` option
- ✅ FileWriter flat mode integration
- ✅ Backward compatibility (nested mode by default)
- ✅ Explicit nested mode specification
- ✅ Error handling for directory creation failures

### Documentation

- Created comprehensive implementation plan: `docs/features/flat/PHASE_4_STEP_4_1.md`
- Includes detailed technical approach, testing strategy, and edge cases

---

## Test Results

```
✅ All 444 tests passing (added 7 new tests)
✅ Coverage: 99.47% (exceeds 99% target)
✅ TypeScript compilation: No errors
✅ Build: Successful
```

---

## Backward Compatibility

✅ **Fully backward compatible** - Default behavior unchanged:
- No `outputStructure` option → nested mode (existing behavior)
- Existing tests continue to pass without modification
- Nested mode code path preserved exactly as before

---

## Data Flow

**Nested Mode (Default)**:
```
outputDir: /blog/_posts
imageDir: /blog/_posts/my-post
imagePathPrefix: .
output: /blog/_posts/my-post/index.md
```

**Flat Mode**:
```
outputDir: /blog/_posts
imageDir: /blog/_images (sibling)
imagePathPrefix: /images
output: /blog/_posts/my-post.md
```

---

## Next Steps

After merging, the following steps remain for complete flat mode support:

- [ ] **Step 4.2**: Update `convertAllPosts` method
  - Create FileWriter with correct mode for `postExists` check
  - Create shared image directory once at start (optimization)

- [ ] **Step 4.3**: Write full end-to-end integration tests

- [ ] **Phase 5**: Add CLI flags (`--flat`, `--image-folder`, `--image-prefix`)

---

## Review Checklist

- [x] All tests passing (444/444)
- [x] Coverage maintained at 99%+
- [x] TypeScript compilation passes
- [x] Build succeeds
- [x] Backward compatibility verified
- [x] Code follows existing patterns
- [x] Comprehensive error handling
- [x] Clear comments explaining logic
- [x] Integration tests cover all scenarios

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/converter.ts [hunk 1]
```diff
@@ -424,36 +496,127 @@ export class Converter extends EventEmitter {
...
+   *
+   * @private
+   */
+  private async convertPostFlat(
+    post: HashnodePost,
+    slug: string,
+    outputDir: string,
+    options: ConversionOptions | undefined,
+    outputStructure: OutputStructure
+  ): Promise<ConvertedPost> {
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply src/converter.ts:618]: Addressed in e05ae9c

File: src/converter.ts [hunk 2]
```diff
@@ -69,13 +75,38 @@ export interface ConverterDependencies {
...
+   * Default dependencies used when not provided via dependency injection.
+   * Follows the same pattern as FileWriter.DEFAULTS for consistency.
+   */
+  private static readonly DEFAULTS: Required<Omit<ConverterDependencies, 'logger' | 'config'>> = {
+    postParser: new PostParser(),
+    markdownTransformer: new MarkdownTransformer(),
+    imageProcessor: new ImageProcessor(),
+    frontmatterGenerator: new FrontmatterGenerator(),
+    fileWriter: new FileWriter(),
+  };
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply src/converter.ts]: Addressed in [9aa996c](https://github.com/alvincrespo/hashnode-content-converter/pull/84/commits/9aa996ccadc58120b0a7e88811cbc1916062ed4e)

File: tests/integration/converter.test.ts [hunk 3]
```diff
@@ -713,6 +720,160 @@ describe('Converter', () => {
...
+        fileWriter: mockFileWriter,
+        config: { outputStructure: { mode: 'flat' } },
+      });
+
+      await flatConverter.convertPost(samplePost, '/output');
+
+      expect(mockImageProcessor.processWithContext).toHaveBeenCalledWith(
+        '# Test Content',
+        expect.objectContaining({
+          imageDir: '/_images',
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply tests/integration/converter.test.ts]: Addressed in a14b00b

File: tests/integration/converter.test.ts [hunk 4]
```diff
@@ -713,6 +720,160 @@ describe('Converter', () => {
...
+        fileWriter: mockFileWriter,
+        config: { outputStructure: { mode: 'flat', imageFolderName: 'assets' } },
+      });
+
+      await flatConverter.convertPost(samplePost, '/src/_posts');
+
+      expect(mockImageProcessor.processWithContext).toHaveBeenCalledWith(
+        expect.any(String),
+        expect.objectContaining({
+          imageDir: '/src/assets',
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply tests/integration/converter.test.ts:797]: Not necessary.

File: docs/features/flat/PHASE_4_STEP_4_1.md [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+From [docs/IMPLEMENTATION_FLAT.md](docs/IMPLEMENTATION_FLAT.md) (lines 764-870):
+
+- Read `outputStructure` from options (default to nested mode for backward compatibility)
+- Calculate image directory based on mode:
+  - **Nested**: `{output}/{slug}/`
+  - **Flat**: `{output}/../{imageFolderName}/` (default: `_images`)
+- Create image directory before processing
+- Use `processWithContext()` for flat mode (with custom `imagePathPrefix`)
+- Use existing `process()` for nested mode (backward compatible)
+- Create FileWriter with appropriate `outputMode`
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply docs/features/flat/PHASE_4_STEP_4_1.md]: Addressed in 6ad0c34

File: docs/features/flat/PHASE_4_STEP_4_1.md [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+├─ Step 1: PostParser.parse(post)
+├─ Step 2: MarkdownTransformer.transform()
+├─ Step 3: Calculate imageDir & imagePathPrefix by mode  ← NEW LOGIC
+│  ├─ Flat:   imageDir = {output}/../_images, prefix = /images
+│  └─ Nested: imageDir = {output}/{slug}, prefix = .
+├─ Step 4: Route to correct ImageProcessor method        ← CONDITIONAL
+│  ├─ Flat:   processWithContext(markdown, { imageDir, imagePathPrefix })
+│  └─ Nested: process(markdown, imageDir)
+├─ Step 5: FrontmatterGenerator.generate()
+└─ Step 6: Create FileWriter with outputMode & write     ← NEW INSTANCE
```
[the remark this answers is not in the document]
Comment (alvincrespo) [reply docs/features/flat/PHASE_4_STEP_4_1.md]: Addressed in 6ad0c34
